bmq-examples/src/main/java/com/bloomberg/bmq/examples/SubscriptionsConsumer.java

```diff
@@ -0,0 +1,322 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Sample application which connects to BlazingMQ broker and consumes messages. Part of the
+ * BlazingMQ tutorial.
+ *
+ * <h2>Interactive Producer/Consumer use case</h2>
+ *
+ * The simplest use case involves a Producer, a Consumer and a Message Broker. The Producer and/or
+ * the Consumer create a named queue in the Broker. The Producer pushes messages into the queue, and
+ * the consumer pulls them. Once a message has been consumed it can be removed from the queue.
+ *
+ * <p>In essence the queue is a mailbox for messages to be sent to a Consumer. It allows decoupling
+ * the processing between the Producer and the Consumer: one of the tasks may be down without
+ * impacting the other, and the Consumer can process messages at a different pace than the Producer.
+ *
+ * <h2>Consumer</h2>
+ *
+ * In a nutshell, this example does the following:
+ *
+ * <ul>
+ *   <li>1. Connects to BlazingMQ framework
+ *   <li>2. Opens queue in read mode
+ *   <li>3. Consumes messages from the queue
+ * </ul>
+ *
+ * <p>In order to consume messages, {@code Consumer} class implements {@link PushMessageHandler}
+ * interface. When new messages are being delivered from the queue, {@link #handlePushMessage}
+ * method is called.
+ *
+ * <p>Also, {@code Consumer} implements {@link SessionEventHandler} and {@link QueueEventHandler}
+ * interfaces in order to be able to handle various events related to {@link Session} and {@link
+ * Queue}.
+ *
+ * <p>Finally, {@code Consumer} implements {@link AutoCloseable} interface. When used in
+ * 'try-with-resources' statement, {@link #close} method will be called automatically after
+ * execution of the block. Thus the connection to BlazingMQ framework will be properly closed.
+ */
+public class SubscriptionsConsumer
+        implements AutoCloseable, SessionEventHandler, QueueEventHandler, PushMessageHandler {
+
+    static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    static final Uri uri = new Uri("bmq://bmq.test.mem.priority/test-queue");
+
+    AbstractSession session;
+    Queue queue;
+
+    public SubscriptionsConsumer() {
+        // Empty: call `run` to start the built Consumer
+    }
+
+    @Override
+    public void handleSessionEvent(SessionEvent event) {
+        // User specified handler for session events
+
+        logger.info("handleSessionEvent: {}", event);
+    }
+
+    @Override
+    public void handleQueueEvent(QueueControlEvent event) {
+        // User specified handler for queue events
+
+        logger.info("handleQueueEvent: {}", event);
+    }
+
+    @Override
+    public void handlePushMessage(PushMessage msg) {
+        // User specified PUSH event handler
+
+        logger.info("*** Got a message from the queue.***");
+        logger.info("GUID: {}", msg.messageGUID());
+        logger.info("CorrelationId: {}", msg.subscription().getCorrelationId().userData());
```

Comment:
I would also add something like here:
https://github.com/bloomberg/blazingmq/blob/7d5fce3199ce6d8330aa33f6e364bd350ad6988e/src/tutorials/subscriptions/consumer.cpp#L426

In order to show how subscriptions can be used later to identify messages